### DISCUSSION TITLE
CrewAI - Agent observability or callbacks

### DISCUSSION THREAD

**Discussion by nethi** [crewAIInc/crewAI#1389] [General]
I am trying to build a general purpose multi-agent chat UI for frameworks like CrewAI, AutoGen etc.  While I have good progress made with AutoGen, I am struggling to get information on agent callbacks with CrewAI. I am looking at step_callbacks but the argument seems to be AgentAction class which doesn't have any info on which agent it is meant for.  Whereas, task callbacks have TaskOutput argument that clearly indicates task that generated the output and the agent it belongs to etc.

It looks like callbacks argument on Agents is removed with some recent changes.

I looked at Phoenix and Langtrace implementations to see how they are instrumenting for crewAI observability - they all seem to monkey patch crewAI to get their job done.

I am wondering shouldn't there be first class support in crewAI for this ?

**Comment by mj-deving** [comment]
The callbacks situation has improved a lot since this was posted. CrewAI moved away from the old `step_callbacks`/`task_callbacks` approach toward a proper event system.

Here's what works now:

**Event Listeners (the modern approach)**

CrewAI has a `BaseEventListener` class with an event bus that gives you granular, per-agent callbacks. This is what you want for a chat UI:

```python
from crewai.utilities.events import BaseEventListener
from crewai.utilities.events.agent_events import (
    AgentExecutionStartedEvent,
    AgentExecutionCompletedEvent,
)
from crewai.utilities.events.task_events import (
    TaskStartedEvent,
    TaskCompletedEvent,
)

class ChatUIListener(BaseEventListener):
    def on_agent_execution_started(self, event: AgentExecutionStartedEvent):
        # event.agent gives you the specific agent
        send_to_ui(f"Agent '{event.agent.role}' started working...")

    def on_agent_execution_completed(self, event: AgentExecutionCompletedEvent):
        send_to_ui(f"Agent '{event.agent.role}' finished: {event.output}")

    def on_task_started(self, event: TaskStartedEvent):
        send_to_ui(f"Task started: {event.task.description[:80]}")

    def on_task_completed(self, event: TaskCompletedEvent):
        send_to_ui(f"Task done: {event.output.raw[:200]}")

# Register before kickoff
listener = ChatUIListener()
crew.kickoff()
```

This gives you exactly what `step_callbacks` never could — which agent is running, what it's doing, and when it finishes.

**OpenTelemetry integration**

If you want structured observability (traces, spans, metrics) rather than raw callbacks, CrewAI has native OTEL support via `opentelemetry-instrumentation-crewai`. Platforms like OpenLIT, Langfuse, and AgentOps all have CrewAI integrations that plug into this.

For a chat UI specifically, the event listener approach is more practical since you need real-time streaming. OTEL is better for post-hoc analysis and dashboards.

The old monkey-patching approach you saw in Phoenix/Langtrace was the only option before the event system existed. You shouldn't need that anymore.
